pkg/services/ngalert/state/manager.go

```diff
@@ -342,33 +379,18 @@ func (st *Manager) staleResultsHandler(ctx context.Context, evaluatedAt time.Tim
 		return stateIsStale(evaluatedAt, s.LastEvaluationTime, alertRule.IntervalSeconds)
 	})
 
-	toDelete := make([]ngModels.AlertInstanceKey, 0)
-
 	for _, s := range staleStates {
 		logger.Info("Detected stale state entry", "cacheID", s.CacheID, "state", s.State, "reason", s.StateReason)
+		oldState := s.State
+		oldReason := s.StateReason
 
-		key, err := s.GetAlertInstanceKey()
-		if err != nil {
-			logger.Error("Unable to get alert instance key to delete it from database. Ignoring", "error", err.Error())
-		} else {
-			toDelete = append(toDelete, key)
-		}
-
-		if s.State == eval.Alerting {
-			oldState := s.State
-			oldReason := s.StateReason
+		s.State = eval.Normal
+		s.StateReason = ngModels.StateReasonMissingSeries
+		s.EndsAt = evaluatedAt
+		s.LastEvaluationTime = evaluatedAt
 
-			s.State = eval.Normal
-			s.StateReason = ngModels.StateReasonMissingSeries
-			s.EndsAt = evaluatedAt
+		if oldState == eval.Alerting {
 			s.Resolved = true
-			s.LastEvaluationTime = evaluatedAt
-			record := StateTransition{
-				State:               s,
-				PreviousState:       oldState,
-				PreviousStateReason: oldReason,
-			}
-
 			// If there is no resolved image for this rule then take one
 			if resolvedImage == nil {
 				image, err := takeImage(ctx, st.imageService, alertRule)
```

Comment:
Should we also find a way to extract the screenshotting? It'd be really nice if this function _only_ cared about handling stale states, and `ProcessEvalResults` handled all the knock-on effects at once, like history/screenshots/etc.

Reply:
This is also technically a database interaction just like `historian` and `instanceStore`.

Reply:
I think this is a great idea. @grobinson-grafana wdyt about moving screenshot logic to ProcessEvalResults?

Reply:
Perhaps like this? https://github.com/grafana/grafana/pull/58476